full/coupler_main.F90

```diff
@@ -1359,7 +1359,7 @@ subroutine coupler_init
     !--- dynamic threading turned off when affinity placement is in use
 !$  call omp_set_dynamic(.FALSE.)
     !--- nested OpenMP enabled for OpenMP concurrent components
-!$  call omp_set_nested(.TRUE.)
+!$  call omp_set_max_active_levels(huge(0)) 
```

Comment:
I think we need to restrict the levels to something other than the number of allowable integers. IIRC, the most we would have used in concurrent radiation was 2.

Reply:
line 1370 below has
```fortran
!$      call fms_affinity_set('ATMOS', use_hyper_thread, atmos_nthreads + radiation_nthreads)
```
Which seems like if  `atmos_nthreads>1` and `radiation_threads>1`, then there could be more than 2.  Or am I misinterpreting this set up?

Reply:
As I think about it, the nesting level needs to be set to "3".

In concurrent radiation, there are a pool of threads granted to radiation and to the a-l-i group.  If one were to simply parcel them out by segmenting based on thread_num, "!$omp master" and "!$omp single" constructs would not be handled correctly within the shared code, basically FMS and FMSCoupler.  In order to get around this and give each thread pool it's own controlling structure, some kludgery/trickery is used.   The top level [parallel region](https://github.com/NOAA-GFDL/FMScoupler/blob/main/full/coupler_main.F90#L780-L788) (level-0) is hard coded to spawn 2 threads where thread 0 is for the a-l-i group and thread 1 is for the radiation to run concurrently.  In this case, we only have one master thread, the one associated with the a-l-i group.  To get around this, each of these threads immediately enters a parallel region spawning a single thread (hard coded) descending into omp nested level-1.  The reason for this is to ensure the [radiation](https://github.com/NOAA-GFDL/FMScoupler/blob/main/full/coupler_main.F90#L897-L903) and the [a-l-i group](https://github.com/NOAA-GFDL/FMScoupler/blob/main/full/coupler_main.F90#L790-L798) each have the concept of a "master" thread for the previously mentioned constructs within the shared code.  With the radiation and a-l-i group each within the level-1 nesting, a call is made to omp_set_num_threads with [radiation_nthreads](https://github.com/NOAA-GFDL/FMScoupler/blob/main/full/coupler_main.F90#L904) or [atmos_nthreads](https://github.com/NOAA-GFDL/FMScoupler/blob/main/full/coupler_main.F90#L799) argument, respectively.

All of the aforementioned parallel regions end prior to the atmospheric system state update.  At this point, all of the threads are pooled together via a call to [omp_set_num_threads](https://github.com/NOAA-GFDL/FMScoupler/blob/main/full/coupler_main.F90#L923) to utilize all available thread resources for the update using the computed fluxes.

Reply:
I don't think setting the large value will allow that many active levels, instead ["The effect of this routine is to set the value of the max-active-levels-var ICV to the value specified in the argument. If the number of active levels requested exceeds the number of active levels of parallelism supported by the implementation, the value of the max-active-levels-var ICV will be set to the number of active levels supported by the implementation."](https://www.openmp.org/spec-html/5.0/openmpsu125.html)  It will then be limited to the number allowed by the implementation.  Is there a reason to restrict it?  I set it up this way because I thought that this was mimicking the old behavior.

Reply:
I would prefer us to set an explicit limit of 3 based on our knowledge of the required levels in order to not get into situations where machine default settings can lead to problems due to ill-defined defaults.

Reply:
After our in person discussion, I changed it to `3`.